backend/cfg/vectorize.ml

```diff
@@ -178,61 +487,828 @@ end = struct
+    let argument_set = reg_array_to_set << Instruction.arguments
+    and affected_set instruction =
+      Reg.Set.union
+        (Instruction.results instruction |> reg_array_to_set)
+        (Instruction.destroyed instruction |> reg_array_to_set)
+    in
+    let arguments_1 = argument_set instruction_1
+    and affected_1 = affected_set instruction_1
+    and arguments_2 = argument_set instruction_2
+    and affected_2 = affected_set instruction_2 in
+    if Reg.Set.disjoint affected_1 affected_2
+       && Reg.Set.disjoint arguments_1 affected_2
+       && Reg.Set.disjoint affected_1 arguments_2
+    then
+      let get_memory_operation instruction =
+        Instruction.Id.Tbl.find_opt t.memory_operations
+          (Instruction.id instruction)
+      in
+      match
+        get_memory_operation instruction_1, get_memory_operation instruction_2
+      with
+      | None, _ | _, None ->
+        Instruction.can_cross_loads_or_stores instruction_1
+        || Instruction.can_cross_loads_or_stores instruction_2
+      | Some memory_operation_1, Some memory_operation_2 -> (
+        match memory_operation_1.op, memory_operation_2.op with
+        | Load, Load -> true
+        | Load, Store | Store, Load | Store, Store ->
+          if Memory_operation.compare_addressing_modes_and_arguments
+               memory_operation_1 memory_operation_2
+             = 0
+          then
+            let check_direct_separation left_memory_operation
+                right_memory_operation =
+              match
+                Memory_operation.offset_of left_memory_operation
+                  right_memory_operation
+              with
+              | None -> false
+              | Some offset ->
+                offset
+                > Cmm.width_of
+                    left_memory_operation.Memory_operation.memory_chunk
+            in
+            check_direct_separation memory_operation_1 memory_operation_2
+            || check_direct_separation memory_operation_2 memory_operation_1
+          else
+            Instruction.Id.Set.is_empty
+              (Instruction.Id.Set.diff memory_operation_1.dependent_allocs
+                 (Instruction.Id.Set.union memory_operation_2.dependent_allocs
+                    memory_operation_2.unsure_allocs))
+            |> not
+            || Instruction.Id.Set.is_empty
+                 (Instruction.Id.Set.diff memory_operation_2.dependent_allocs
+                    (Instruction.Id.Set.union
+                       memory_operation_1.dependent_allocs
+                       memory_operation_1.unsure_allocs))
+               |> not)
+    else false
+
+  let can_cross_lists t instructions1 instructions2 =
+    let can_cross_list instruction1 =
+      List.fold_left
+        (fun can instruction2 -> can && can_cross t instruction1 instruction2)
+        true instructions2
+    in
+    List.fold_left
+      (fun can instruction1 -> can && can_cross_list instruction1)
+      true instructions1
```

Comment:
(Can possibly be simplified by
using `List.for_all`.)

Reply:
done